File: typescript/src/mapInvoicePayment.ts
import type { StripeReceivedHookEvent } from 'subscrio';

export interface ExtractedInvoicePayment {
  amountPaid: number;
  currency: string;
  stripeInvoiceId: string;
  stripeEventId: string | null;
  stripeSubscriptionId: string;
}

function asId(value: unknown): string | undefined {
  if (typeof value === 'string' && value.length > 0) return value;
  if (value && typeof value === 'object' && 'id' in value) {
    const id = (value as { id?: unknown }).id;
    if (typeof id === 'string' && id.length > 0) return id;
  }
  return undefined;
}

function asAmount(value: unknown): number {
  if (typeof value === 'number' && Number.isFinite(value)) return Math.trunc(value);
  if (typeof value === 'string' && value.length > 0) {
    const n = Number(value);
    if (Number.isFinite(n)) return Math.trunc(n);
  }
  return 0;
}

/**
 * Pull payment fields from a stripe.received.after hook when the event is invoice.payment_succeeded.
 * Returns null for other event types or invoices without a Stripe subscription / invoice id.
 */
export function mapInvoicePayment(event: StripeReceivedHookEvent): ExtractedInvoicePayment | null {
  if (event.data?.type !== 'invoice.payment_succeeded') {
    return null;
  }

  const obj = event.data.data?.object as unknown as Record<string, unknown> | undefined;
  if (!obj) {
    return null;
  }

  const stripeInvoiceId = asId(obj.id);
  if (!stripeInvoiceId) {
    return null;
  }

  const parent = obj.parent as Record<string, unknown> | undefined;
  const subscriptionDetails = parent?.subscription_details as Record<string, unknown> | undefined;
  const stripeSubscriptionId =
    asId(obj.subscription) ?? asId(subscriptionDetails?.subscription);
  if (!stripeSubscriptionId) {
    return null;
  }

  return {
    amountPaid: asAmount(obj.amount_paid),
    currency: typeof obj.currency === 'string' ? obj.currency : '',
    stripeInvoiceId,
    stripeEventId: typeof event.data.id === 'string' ? event.data.id : null,
    stripeSubscriptionId,
  };
}

File: dotnet/src/Mapping/MapInvoicePayment.cs
using Stripe;
using Subscrio.Core.Application.Hooks;

namespace Subscrio.Payments.Mapping;

public sealed class ExtractedInvoicePayment
{
    public required int AmountPaid { get; init; }
    public required string Currency { get; init; }
    public required string StripeInvoiceId { get; init; }
    public string? StripeEventId { get; init; }
    public required string StripeSubscriptionId { get; init; }
}

public static class MapInvoicePayment
{
    public static ExtractedInvoicePayment? FromEvent(StripeReceivedHookEvent evt)
    {
        if (evt.Data?.Type != EventTypes.InvoicePaymentSucceeded)
            return null;

        if (evt.Data.Data?.Object is not Invoice invoice)
            return null;

        if (string.IsNullOrEmpty(invoice.Id))
            return null;

        var stripeSubscriptionId = AsId(invoice.Parent?.SubscriptionDetails?.SubscriptionId)
            ?? AsId(invoice.Parent?.SubscriptionDetails?.Subscription?.Id);
        if (string.IsNullOrEmpty(stripeSubscriptionId))
            return null;

        var amountPaid = invoice.AmountPaid > int.MaxValue
            ? int.MaxValue
            : (int)invoice.AmountPaid;

        return new ExtractedInvoicePayment
        {
            AmountPaid = amountPaid,
            Currency = invoice.Currency ?? string.Empty,
            StripeInvoiceId = invoice.Id,
            StripeEventId = string.IsNullOrEmpty(evt.Data.Id) ? null : evt.Data.Id,
            StripeSubscriptionId = stripeSubscriptionId
        };
    }

    private static string? AsId(string? value) =>
        string.IsNullOrEmpty(value) ? null : value;
}
